### PR TITLE
Expand HashedModuleIdsPlugin to assign hash based IDs to DelegatedMod…

### PR DESCRIPTION
**Pull request by matthewmeyer**
…ules in addition to NormalModules

**What kind of change does this PR introduce?**

Enhancement, expected behavior.

**Did you add tests for your changes?**

Yes, configCases/delegated-hash

**If relevant, link to documentation update:**

N/A

**Summary**

While using the HashedModuleIdsPlugin to create stable IDs, DelegatedModules like those generated from DllReferencePlugin were still unstable and were changing between builds. Updated HashedModuleIdsPlugin to generate hashes for DelegatedModules, and DelegatedModule to support non-number moduleIds.

**Does this PR introduce a breaking change?**

No

**Other information**

### CODE REVIEW

**Comment by sainthardaway** [comment]
:+1:

**Comment by manekovskiy** [comment]
In my opinion, support of non-number module ids by DelegatedModule is essential. I was surprised when found that webpack does not this. This PR would be a great addition. Thanks 👍

File: lib/DelegatedModule.js [hunk 1]
```diff
@@ -57,7 +57,11 @@ class DelegatedModule extends Module {
 		if(!sourceModule) {
 			str = WebpackMissingModule.moduleCode(this.sourceRequest);
 		} else {
-			str = `module.exports = (__webpack_require__(${sourceModule.id}))`;
+			if(typeof sourceModule.id === "string") {
+				str = `module.exports = (__webpack_require__("${sourceModule.id}"))`;
```
Comment (sokra) [inline lib/DelegatedModule.js]: use JSON.stringify instead

File: lib/HashedModuleIdsPlugin.js [hunk 2]
```diff
@@ -20,10 +20,16 @@ class HashedModuleIdsPlugin {
 			const usedIds = new Set();
 			compilation.plugin("before-module-ids", (modules) => {
 				modules.forEach((module) => {
-					if(module.id === null && module.libIdent) {
-						const id = module.libIdent({
-							context: this.options.context || compiler.options.context
-						});
+					if(module.id === null) {
```
Comment (sokra) [inline lib/HashedModuleIdsPlugin.js]: add a libIdent method to DelegatedModule instead

File: lib/DelegatedModule.js (hunk 1 above)
Comment (matthewmeyer) [inline lib/DelegatedModule.js]: Thank you, I have updated as requested.

File: lib/HashedModuleIdsPlugin.js (hunk 2 above)
Comment (matthewmeyer) [inline lib/HashedModuleIdsPlugin.js]: Added libIdent method to DelegatedModule and ExternalModule

**Comment by matthewmeyer** [comment]
My last commit only added new tests, but I failed different tests on CI that I didn't touch. Is there a way to rerun the CI (looks like the tests that failed were timeouts on a watcher)?

File: lib/DelegatedModule.js [hunk 3]
```diff
@@ -57,7 +68,7 @@ class DelegatedModule extends Module {
 		if(!sourceModule) {
 			str = WebpackMissingModule.moduleCode(this.sourceRequest);
 		} else {
-			str = `module.exports = (__webpack_require__(${sourceModule.id}))`;
+			str = `module.exports = (__webpack_require__(${JSON.stringify(sourceModule.id)}))`;
```
Comment (sokra) [inline lib/DelegatedModule.js]: Good find 😄

File: lib/DelegatedModule.js [hunk 4]
```diff
@@ -23,6 +24,16 @@ class DelegatedModule extends Module {
...
 
+	contextify(context, request) {
+		return request.split("!").map(function(r) {
+			let rp = path.relative(context, r).replace(/\\/g, "/");
+			return rp.indexOf("../") !== 0 ? "./" + rp : rp;
+		}).join("!");
+	}
+
+	libIdent(options) {
+		return this.contextify(options.context, this.userRequest);
```
Comment (sokra) [inline lib/DelegatedModule.js]: contextify is not needed here als userRequest is already an lib ident

File: lib/ExternalModule.js [hunk 5]
```diff
@@ -18,6 +18,10 @@ class ExternalModule extends Module {
 		this.external = true;
 	}
 
+	libIdent() {
+		return this.request;
```
Comment (sokra) [inline lib/ExternalModule.js]: also use userRequest here.

this is not passed to the module yet:
add a userRequest constructor argumunt and pass it from the ExternalsPlugin

File: lib/DelegatedModule.js (hunk 4 above)
Comment (matthewmeyer) [inline lib/DelegatedModule.js]: When I remove contextify I get test case failures from: "ConfigTestCases dll-plugin 1-use-dll should load an harmony module from dll (star export)"

It looks like the test case has two modules './e' one in the DllModule and the other locally. They have the same name so they conflict in the bundle.

**Comment by matthewmeyer** [comment]
Do I need to resubmit this, how does this move forward?

**Comment by sokra** [comment]
@matthewmeyer I saw it (they are marked with a `PR: review-outdated` label), but I still need to think about the contextify thing....

**Comment by sokra** [comment]
Thanks

**Pull request merged**